packages/custom-url/src/Infrastructure/Sulu/EventSubscriber/CacheInvalidationSubscriber.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\CustomUrl\Infrastructure\Sulu\EventSubscriber;
+
+use Sulu\Bundle\HttpCacheBundle\Cache\CacheManager;
+use Sulu\Bundle\PageBundle\Document\RouteDocument;
+use Sulu\Component\DocumentManager\Behavior\Mapping\UuidBehavior;
+use Sulu\Component\DocumentManager\Event\PublishEvent;
+use Sulu\Component\DocumentManager\Event\RemoveEvent;
+use Sulu\Component\DocumentManager\Events;
+use Sulu\CustomUrl\Domain\Model\CustomUrlInterface;
+use Sulu\CustomUrl\Domain\Model\CustomUrlRoute;
+use Sulu\CustomUrl\Infrastructure\Repository\CustomUrlRepositoryInterface;
+use Sulu\CustomUrl\Infrastructure\Repository\CustomUrlRouteRepositoryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * When a content has changed or been removed we need to invalidate all custom url routes that point to the content
+ * When a custom route has been removed we need to invalidate all the caches too.
+ */
+class CacheInvalidationSubscriber implements EventSubscriberInterface
```

Comment:
This probably has to be split into subscriber for document updates and subscriber for customurl entity updates.